companion/src/firmwares/opentx/opentxinterface.cpp

```diff
@@ -222,33 +237,72 @@ unsigned long OpenTxEepromInterface::loadxml(RadioData &radioData, QDomDocument
   return errors.to_ulong();
 }
 
-int OpenTxEepromInterface::loadFile(RadioData & radioData, const QString & filename)
+QList<OpenTxEepromInterface *> opentxEEpromInterfaces;
+void registerOpenTxEEpromInterface(BoardEnum board)
+{
+  OpenTxEepromInterface * interface = new OpenTxEepromInterface(board);
+  opentxEEpromInterfaces.push_back(interface);
+  eepromInterfaces.push_back(interface);
+}
+
+void registerOpenTxEEpromInterfaces()
+{
+  registerOpenTxEEpromInterface(BOARD_STOCK);
+  registerOpenTxEEpromInterface(BOARD_M128);
+  registerOpenTxEEpromInterface(BOARD_GRUVIN9X);
+  registerOpenTxEEpromInterface(BOARD_SKY9X);
+  registerOpenTxEEpromInterface(BOARD_9XRPRO);
+  registerOpenTxEEpromInterface(BOARD_TARANIS_X9D);
+  registerOpenTxEEpromInterface(BOARD_TARANIS_X9DP);
+  registerOpenTxEEpromInterface(BOARD_TARANIS_X9E);
+  registerOpenTxEEpromInterface(BOARD_HORUS);
+}
+
+template <class T, class M>
+bool loadFromByteArray(T & dest, const QByteArray & data)
+{
+  foreach(OpenTxEepromInterface * eepromInterface, opentxEEpromInterfaces) {
+    if (eepromInterface->loadFromByteArray<T, M>(dest, data)) {
+      return true;
+    }
+  }
+  
+  return false;
+};
+
+bool loadFile(RadioData & radioData, const QString & filename)
 {
   StorageSdcard storage;
 
   storage.read(filename);
 
   // Radio settings
   qDebug() << "Radio settings:" << storage.radio.size();
-  loadFromByteArray<GeneralSettings, OpenTxGeneralData>(radioData.generalSettings, storage.radio);
+  if (!loadFromByteArray<GeneralSettings, OpenTxGeneralData>(radioData.generalSettings, storage.radio))
+    return false;
 
   // Models
   int modelIndex = 0;
   QString modelList = QString(storage.modelList);
   QList<QByteArray> lines = storage.modelList.split('\n');
-  QString category = QObject::tr("Unknown");
+  int categoryIndex = -1;
   foreach (const QByteArray & line, lines) {
     if (!line.isEmpty()) {
       if (line.startsWith('[') && line.endsWith(']')) {
-        category = line.mid(1, line.size() - 2);
+        QString name = line.mid(1, line.size() - 2);
+        CategoryData category;
+        strncpy(category.name, name.toStdString().c_str(), sizeof(CategoryData::name));
+        radioData.categories.push_back(category);
+        categoryIndex++;
       }
       else {
         qDebug() << "Loading" << line;
         foreach (const ModelFile &model, storage.models) {
           if (line == model.filename) {
-            loadFromByteArray<ModelData, OpenTxModelData>(radioData.models[modelIndex], model.data);
+            if (!loadFromByteArray<ModelData, OpenTxModelData>(radioData.models[modelIndex], model.data))
+              return false;
```

Comment:
this method also allows us to store all EEPROM files in .otx format, this is not too bad! but sure there is a possible optimization because this code tries all possible variants for EACH model, which is a nonsense. not an easy one